Builder escapes worktree and implements wrong issue under degraded sessions

## Problem

During the shepherd run for #2609, the builder agent:

1. **Escaped the worktree** — modified files on main instead of in `.loom/worktrees/issue-2609`
2. **Implemented the wrong issue** — produced changes for #2603 (post-curator blocked label check) instead of #2609 (phase-contract-failure comment noise)

The builder log showed rate limit warnings (87% weekly) and repeated "Crystallizing..." loops, suggesting a degraded session where the LLM lost context about both *where* to work and *what* to work on.

The shepherd's validation correctly detected the escape after the fact:
```
[WARN] Builder may have escaped worktree for issue #2609 — main branch has 1 uncommitted file(s)
[ERROR] builder phase validation failed: ... worktree exists (branch=feature/issue-2609, commits_ahead=0, uncommitted=none); remote branch missing; no PR; labels=[loom:curated]; WARNING: main branch dirty (1 NEW files)
```

But by then the run was wasted.

## Impact

- Entire shepherd run is wasted (claim, curator skip, approval gate, builder spawn, validation, cleanup)
- Incorrect changes left on main branch require manual cleanup
- If a previous attempt also left a stale worktree/branch, the next attempt has to deal with that too (as happened here — the worktree had a commit from a prior failed attempt with a different approach)

## Observed During

Shepherd run for #2609 (`/shepherd 2609 -m`). Builder log: `.loom/logs/loom-builder-issue-2609.log`

## Suggested Improvements

- **Pre-flight worktree anchor**: Before spawning the builder, verify `pwd` is inside the worktree and inject a stronger prompt reminder about the working directory
- **Post-builder dirty-main check before validation**: If main has new dirty files after builder exits, immediately classify as worktree escape and skip mechanical recovery (which can't succeed anyway)
- **Wrong-issue detection**: After builder exits, check if any commits/changes reference a different issue number than the one assigned — flag as confused session

--- a/loom-tools/src/loom_tools/shepherd/phases/builder.py
+++ b/loom-tools/src/loom_tools/shepherd/phases/builder.py
@@ -374,6 +374,27 @@ class BuilderPhase:
         # Create marker to prevent premature cleanup
         self._create_worktree_marker(ctx)
 
+        # Pre-flight worktree anchor verification (issue #2630):
+        # Confirm the worktree is a valid git directory before spawning the
+        # builder.  This provides a clear audit trail.  We warn instead of
+        # failing because the builder would fail naturally on git operations
+        # if the worktree is invalid.
+        if ctx.worktree_path and ctx.worktree_path.is_dir():
+            anchor_check = subprocess.run(
+                ["git", "-C", str(ctx.worktree_path), "rev-parse", "--git-dir"],
+                capture_output=True, text=True, check=False,
+            )
+            if anchor_check.returncode != 0:
+                log_warning(
+                    f"Worktree at {ctx.worktree_path} may not be a valid "
+                    f"git directory — builder may fail"
+                )
+            else:
+                log_info(
+                    f"Worktree anchor verified: builder will work in "
+                    f"{ctx.worktree_path} (branch={ctx.worktree_path.name})"
+                )
+
         # Snapshot main's dirty state before spawning the builder so we can
         # distinguish pre-existing dirt from actual worktree escapes later.
         self._main_dirty_baseline = self._snapshot_main_dirty(ctx)
@@ -548,6 +569,15 @@ class BuilderPhase:
                 phase_name="builder",
                 data={"exit_code": exit_code, "diagnostics": diag},
             )
+
+        # Early worktree escape detection (issue #2630):
+        # Check for dirty main immediately after builder exits — before
+        # entering the validation/completion loop.  When the builder
+        # escapes the worktree and modifies main instead, completion
+        # retries are futile (they operate on the empty worktree).
+        escape = self._detect_worktree_escape(ctx)
+        if escape is not None:
+            return escape
 
         # Run test verification in worktree (unless explicitly skipped)
         if skip_test_verification:
@@ -2825,6 +2855,159 @@ class BuilderPhase:
             )
         return set()
 
+    def _detect_worktree_escape(
+        self, ctx: ShepherdContext
+    ) -> PhaseResult | None:
+        """Detect worktree escape early — before the validation/completion loop.
+
+        When the builder escapes the worktree and modifies main instead,
+        two signals are present simultaneously:
+          1. Main has NEW dirty files (compared to the pre-builder baseline).
+          2. The worktree is clean (0 commits ahead, no uncommitted changes).
+
+        When both hold, completion retries are futile — they operate on the
+        empty worktree — so we short-circuit to FAILED immediately.
+
+        Also detects wrong-issue confusion: if the worktree has commits but
+        they reference a different issue number than the one assigned.
+
+        Returns:
+            PhaseResult if escape or wrong-issue detected, else None.
+
+        See issue #2630.
+        """
+        wt = ctx.worktree_path
+        if not wt or not wt.is_dir():
+            return None
+
+        # --- Check for dirty main (worktree escape) --------------------------
+        new_dirty = self._get_new_main_dirty_files(ctx)
+        if new_dirty:
+            # Check if worktree is empty (no real work there)
+            wt_status = subprocess.run(
+                ["git", "-C", str(wt), "status", "--porcelain"],
+                capture_output=True, text=True, check=False,
+            )
+            wt_uncommitted = bool(
+                wt_status.returncode == 0 and wt_status.stdout.strip()
+            )
+            log_res = subprocess.run(
+                ["git", "-C", str(wt), "log", "--oneline", "main..HEAD"],
+                capture_output=True, text=True, check=False,
+            )
+            commits_ahead = (
+                len([l for l in log_res.stdout.splitlines() if l])
+                if log_res.returncode == 0 and log_res.stdout.strip()
+                else 0
+            )
+
+            if commits_ahead == 0 and not wt_uncommitted:
+                # Classic escape: builder worked on main, worktree is untouched
+                dirty_preview = new_dirty[:5]
+                log_error(
+                    f"Builder escaped worktree for issue #{ctx.config.issue} — "
+                    f"main branch has {len(new_dirty)} new dirty file(s): "
+                    f"{dirty_preview}"
+                )
+                self._cleanup_stale_worktree(ctx)
+                return PhaseResult(
+                    status=PhaseStatus.FAILED,
+                    message=(
+                        f"builder escaped worktree: main branch has "
+                        f"{len(new_dirty)} new dirty file(s) but worktree "
+                        f"is clean (0 commits, no uncommitted changes)"
+                    ),
+                    phase_name="builder",
+                    data={
+                        "worktree_escape": True,
+                        "main_dirty_files": new_dirty[:10],
+                    },
+                )
+
+        # --- Check for wrong-issue confusion ----------------------------------
+        wrong_issue = self._detect_wrong_issue(ctx)
+        if wrong_issue is not None:
+            issue_refs, commit_messages = wrong_issue
+            log_error(
+                f"Builder confused: commits in worktree for issue "
+                f"#{ctx.config.issue} reference other issue(s): "
+                f"{issue_refs}"
+            )
+            self._cleanup_stale_worktree(ctx)
+            return PhaseResult(
+                status=PhaseStatus.FAILED,
+                message=(
+                    f"builder confused session: commits reference "
+                    f"issue(s) {issue_refs} instead of #{ctx.config.issue}"
+                ),
+                phase_name="builder",
+                data={
+                    "wrong_issue": True,
+                    "referenced_issues": sorted(issue_refs),
+                    "commit_messages": commit_messages,
+                },
+            )
+
+        return None
+
+    def _get_new_main_dirty_files(self, ctx: ShepherdContext) -> list[str]:
+        """Return list of NEW dirty files on main since baseline snapshot."""
+        main_status = subprocess.run(
+            ["git", "-C", str(ctx.repo_root), "status", "--porcelain"],
+            capture_output=True, text=True, check=False,
+        )
+        if not (main_status.returncode == 0 and main_status.stdout.strip()):
+            return []
+
+        current = [line for line in main_status.stdout.splitlines() if line]
+        if self._main_dirty_baseline is not None:
+            return [f for f in current if f not in self._main_dirty_baseline]
+        return current
+
+    def _detect_wrong_issue(
+        self, ctx: ShepherdContext
+    ) -> tuple[set[int], list[str]] | None:
+        """Check if worktree commits reference a different issue number.
+
+        Returns (wrong_issue_numbers, commit_messages) if wrong-issue
+        detected, else None.
+        """
+        wt = ctx.worktree_path
+        if not wt or not wt.is_dir():
+            return None
+
+        log_res = subprocess.run(
+            ["git", "-C", str(wt), "log", "--format=%s", "main..HEAD"],
+            capture_output=True, text=True, check=False,
+        )
+        if log_res.returncode != 0 or not log_res.stdout.strip():
+            return None
+
+        commit_messages = [
+            line for line in log_res.stdout.splitlines() if line
+        ]
+        if not commit_messages:
+            return None
+
+        # Find all issue references (#NNN) in commit messages
+        assigned = ctx.config.issue
+        other_issues: set[int] = set()
+        for msg in commit_messages:
+            refs = re.findall(r"#(\d+)", msg)
+            for ref in refs:
+                num = int(ref)
+                if num != assigned and num > 0:
+                    other_issues.add(num)
+
+        # Only flag when commits reference OTHER issues but NOT the assigned one
+        references_assigned = any(
+            f"#{assigned}" in msg for msg in commit_messages
+        )
+        if other_issues and not references_assigned:
+            return other_issues, commit_messages
+
+        return None
+
     def _gather_diagnostics(self, ctx: ShepherdContext) -> dict[str, Any]:
         """Collect diagnostic info about the builder environment.
 
@@ -3080,6 +3263,16 @@ class BuilderPhase:
         diag["main_dirty_file_count"] = len(new_dirty_files)
         diag["main_dirty_files"] = new_dirty_files[:10]  # Cap for readability
 
+        # -- Wrong-issue detection (issue #2630) ----------------------------
+        wrong = self._detect_wrong_issue(ctx)
+        if wrong is not None:
+            wrong_issues, wrong_msgs = wrong
+            diag["wrong_issue_refs"] = sorted(wrong_issues)
+            diag["wrong_issue_commit_messages"] = wrong_msgs
+        else:
+            diag["wrong_issue_refs"] = []
+            diag["wrong_issue_commit_messages"] = []
+
         # -- Human-readable summary -----------------------------------------
         parts: list[str] = []
         # Surface root cause errors from the log at the front of the
@@ -3128,6 +3321,11 @@ class BuilderPhase:
         elif main_dirty_files and not new_dirty_files:
             parts.append(
                 f"main branch dirty ({len(main_dirty_files)} pre-existing files, ignored)"
+            )
+        if diag.get("wrong_issue_refs"):
+            parts.append(
+                f"WARNING: commits reference wrong issue(s): "
+                f"{diag['wrong_issue_refs']}"
             )
         parts.append(f"log={diag['log_file']}")
         diag["summary"] = "; ".join(parts)

--- a/loom-tools/tests/shepherd/test_builder_escape_detection.py
+++ b/loom-tools/tests/shepherd/test_builder_escape_detection.py
@@ -1,0 +1,305 @@
+"""Tests for builder worktree escape and wrong-issue detection (issue #2630).
+
+Tests the early detection of:
+1. Builder escaping worktree and modifying main instead
+2. Builder commits referencing a different issue number
+3. Pre-flight worktree anchor verification
+"""
+
+from __future__ import annotations
+
+import subprocess
+from pathlib import Path
+from unittest.mock import MagicMock, PropertyMock, patch
+
+import pytest
+
+from loom_tools.shepherd.config import ShepherdConfig
+from loom_tools.shepherd.context import ShepherdContext
+from loom_tools.shepherd.phases import BuilderPhase, PhaseStatus
+
+
+def _make_mock_path(is_dir: bool = True, name: str = "issue-42") -> MagicMock:
+    """Create a MagicMock that behaves like a Path with is_dir()."""
+    mock_path = MagicMock()
+    mock_path.is_dir.return_value = is_dir
+    mock_path.name = name
+    mock_path.__str__ = lambda self: f"/fake/repo/.loom/worktrees/{name}"
+    mock_path.__truediv__ = lambda self, other: MagicMock()
+    return mock_path
+
+
+@pytest.fixture
+def mock_context() -> MagicMock:
+    """Create a mock ShepherdContext."""
+    ctx = MagicMock(spec=ShepherdContext)
+    ctx.config = ShepherdConfig(issue=42)
+    ctx.repo_root = Path("/fake/repo")
+    ctx.scripts_dir = Path("/fake/repo/.loom/scripts")
+    ctx.worktree_path = _make_mock_path(is_dir=True)
+    ctx.pr_number = None
+    ctx.label_cache = MagicMock()
+    return ctx
+
+
+def _make_run_result(
+    returncode: int = 0,
+    stdout: str = "",
+    stderr: str = "",
+) -> subprocess.CompletedProcess[str]:
+    return subprocess.CompletedProcess(
+        args=[], returncode=returncode, stdout=stdout, stderr=stderr
+    )
+
+
+class TestGetNewMainDirtyFiles:
+    """Test _get_new_main_dirty_files filtering."""
+
+    def test_no_dirty_files(self, mock_context: MagicMock) -> None:
+        builder = BuilderPhase()
+        builder._main_dirty_baseline = set()
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run",
+            return_value=_make_run_result(stdout=""),
+        ):
+            result = builder._get_new_main_dirty_files(mock_context)
+        assert result == []
+
+    def test_all_new_dirty_files(self, mock_context: MagicMock) -> None:
+        builder = BuilderPhase()
+        builder._main_dirty_baseline = set()
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run",
+            return_value=_make_run_result(stdout="?? new_file.py\nM changed.py\n"),
+        ):
+            result = builder._get_new_main_dirty_files(mock_context)
+        assert len(result) == 2
+
+    def test_filters_preexisting_dirty_files(self, mock_context: MagicMock) -> None:
+        builder = BuilderPhase()
+        builder._main_dirty_baseline = {"M preexisting.py"}
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run",
+            return_value=_make_run_result(
+                stdout="M preexisting.py\n?? new_file.py\n"
+            ),
+        ):
+            result = builder._get_new_main_dirty_files(mock_context)
+        assert result == ["?? new_file.py"]
+
+    def test_no_baseline_treats_all_as_new(self, mock_context: MagicMock) -> None:
+        builder = BuilderPhase()
+        builder._main_dirty_baseline = None
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run",
+            return_value=_make_run_result(stdout="M file.py\n"),
+        ):
+            result = builder._get_new_main_dirty_files(mock_context)
+        assert result == ["M file.py"]
+
+
+class TestDetectWrongIssue:
+    """Test _detect_wrong_issue commit message analysis."""
+
+    def test_no_commits(self, mock_context: MagicMock) -> None:
+        builder = BuilderPhase()
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run",
+            return_value=_make_run_result(stdout=""),
+        ):
+            result = builder._detect_wrong_issue(mock_context)
+        assert result is None
+
+    def test_commits_reference_correct_issue(self, mock_context: MagicMock) -> None:
+        builder = BuilderPhase()
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run",
+            return_value=_make_run_result(stdout="fix: handle edge case (#42)\n"),
+        ):
+            result = builder._detect_wrong_issue(mock_context)
+        assert result is None
+
+    def test_commits_reference_wrong_issue_only(self, mock_context: MagicMock) -> None:
+        builder = BuilderPhase()
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run",
+            return_value=_make_run_result(
+                stdout="fix: handle post-curator blocked label (#603)\n"
+            ),
+        ):
+            result = builder._detect_wrong_issue(mock_context)
+        assert result is not None
+        wrong_issues, messages = result
+        assert 603 in wrong_issues
+        assert 42 not in wrong_issues
+
+    def test_commits_reference_both_issues_not_flagged(
+        self, mock_context: MagicMock
+    ) -> None:
+        """When commits reference both the assigned AND other issues, don't flag."""
+        builder = BuilderPhase()
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run",
+            return_value=_make_run_result(
+                stdout="fix: resolve #42 (related to #603)\n"
+            ),
+        ):
+            result = builder._detect_wrong_issue(mock_context)
+        assert result is None
+
+    def test_multiple_wrong_issues(self, mock_context: MagicMock) -> None:
+        builder = BuilderPhase()
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run",
+            return_value=_make_run_result(
+                stdout="fix: handle #100 and #200\nchore: cleanup for #300\n"
+            ),
+        ):
+            result = builder._detect_wrong_issue(mock_context)
+        assert result is not None
+        wrong_issues, messages = result
+        assert wrong_issues == {100, 200, 300}
+
+    def test_no_worktree(self, mock_context: MagicMock) -> None:
+        builder = BuilderPhase()
+        mock_context.worktree_path = None
+        result = builder._detect_wrong_issue(mock_context)
+        assert result is None
+
+    def test_no_issue_refs_in_commits(self, mock_context: MagicMock) -> None:
+        """Commits without any issue references should not flag."""
+        builder = BuilderPhase()
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run",
+            return_value=_make_run_result(
+                stdout="fix: handle edge case\nchore: update tests\n"
+            ),
+        ):
+            result = builder._detect_wrong_issue(mock_context)
+        assert result is None
+
+
+class TestDetectWorktreeEscape:
+    """Test _detect_worktree_escape early detection."""
+
+    def test_no_escape_clean_main(self, mock_context: MagicMock) -> None:
+        """No escape when main is clean."""
+        builder = BuilderPhase()
+        builder._main_dirty_baseline = set()
+
+        with patch.object(
+            builder, "_get_new_main_dirty_files", return_value=[]
+        ), patch.object(builder, "_detect_wrong_issue", return_value=None):
+            result = builder._detect_worktree_escape(mock_context)
+        assert result is None
+
+    def test_escape_dirty_main_clean_worktree(self, mock_context: MagicMock) -> None:
+        """Escape detected: main is dirty, worktree is clean."""
+        builder = BuilderPhase()
+        builder._main_dirty_baseline = set()
+
+        with (
+            patch.object(
+                builder,
+                "_get_new_main_dirty_files",
+                return_value=["?? new_file.py"],
+            ),
+            patch(
+                "loom_tools.shepherd.phases.builder.subprocess.run",
+                side_effect=[
+                    # git status --porcelain (worktree) - clean
+                    _make_run_result(stdout=""),
+                    # git log --oneline main..HEAD (worktree) - no commits
+                    _make_run_result(stdout=""),
+                ],
+            ),
+            patch.object(builder, "_cleanup_stale_worktree"),
+        ):
+            result = builder._detect_worktree_escape(mock_context)
+
+        assert result is not None
+        assert result.status == PhaseStatus.FAILED
+        assert result.data["worktree_escape"] is True
+
+    def test_no_escape_when_worktree_has_commits(
+        self, mock_context: MagicMock
+    ) -> None:
+        """Not an escape when worktree has commits (builder worked in both)."""
+        builder = BuilderPhase()
+        builder._main_dirty_baseline = set()
+
+        with (
+            patch.object(
+                builder,
+                "_get_new_main_dirty_files",
+                return_value=["?? leaked_file.py"],
+            ),
+            patch(
+                "loom_tools.shepherd.phases.builder.subprocess.run",
+                side_effect=[
+                    # git status --porcelain (worktree) - clean
+                    _make_run_result(stdout=""),
+                    # git log --oneline main..HEAD (worktree) - has commits
+                    _make_run_result(stdout="abc1234 feat: implement feature\n"),
+                ],
+            ),
+            patch.object(builder, "_detect_wrong_issue", return_value=None),
+        ):
+            result = builder._detect_worktree_escape(mock_context)
+
+        # Not flagged as escape because worktree has commits
+        assert result is None
+
+    def test_wrong_issue_detected(self, mock_context: MagicMock) -> None:
+        """Wrong-issue confusion detected via commit messages."""
+        builder = BuilderPhase()
+        builder._main_dirty_baseline = set()
+
+        with (
+            patch.object(builder, "_get_new_main_dirty_files", return_value=[]),
+            patch.object(
+                builder,
+                "_detect_wrong_issue",
+                return_value=({603}, ["fix: handle #603"]),
+            ),
+            patch.object(builder, "_cleanup_stale_worktree"),
+        ):
+            result = builder._detect_worktree_escape(mock_context)
+
+        assert result is not None
+        assert result.status == PhaseStatus.FAILED
+        assert result.data["wrong_issue"] is True
+        assert 603 in result.data["referenced_issues"]
+
+    def test_no_worktree_returns_none(self, mock_context: MagicMock) -> None:
+        """No check needed when worktree doesn't exist."""
+        builder = BuilderPhase()
+        mock_context.worktree_path = _make_mock_path(is_dir=False)
+        result = builder._detect_worktree_escape(mock_context)
+        assert result is None
+
+    def test_escape_message_includes_file_count(
+        self, mock_context: MagicMock
+    ) -> None:
+        """Error message should include the number of dirty files."""
+        builder = BuilderPhase()
+        builder._main_dirty_baseline = set()
+
+        dirty_files = ["?? file1.py", "?? file2.py", "M file3.py"]
+        with (
+            patch.object(
+                builder, "_get_new_main_dirty_files", return_value=dirty_files
+            ),
+            patch(
+                "loom_tools.shepherd.phases.builder.subprocess.run",
+                side_effect=[
+                    _make_run_result(stdout=""),
+                    _make_run_result(stdout=""),
+                ],
+            ),
+            patch.object(builder, "_cleanup_stale_worktree"),
+        ):
+            result = builder._detect_worktree_escape(mock_context)
+
+        assert result is not None
+        assert "3 new dirty file(s)" in result.message

